File: solutions/cpp/perfect-numbers/2/perfect_numbers.cpp
#include "perfect_numbers.h"
#include <stdexcept>

namespace perfect_numbers {

classification classify(int num){
    int sum = 0;
    if (num<=0) 
        throw std::domain_error("N must be >0.");
    for(int k=1; k<num; k++){
        if (num % k == 0)
            sum += k;}
    if(sum == num)
        return classification::perfect;
    else if (sum > num)
        return classification::abundant;
    else // (sum < num)
        return classification::deficient;
        
}



}  // namespace perfect_numbers

File: solutions/fortran/perfect-numbers/2/perfect_numbers.f90

module perfect_numbers
  implicit none

contains

integer function aliquot_sum(n) result(aliquot)
  integer, intent(in) :: n
  integer :: i
  integer, dimension(n / 2)  :: seq

  seq = [(i, i=1, n / 2)]
  aliquot = sum(seq, mask = mod(n,seq) == 0)
end function aliquot_sum

character(len=9) function classify(num)
    integer, intent(in) :: num
    if (num>0) then
       select case(aliquot_sum(num) - num)
        case (0) 
            classify = "perfect"
       case (1:) 
            classify = "abundant"
       case default
            classify = "deficient"
      end select
    else
      classify = "ERROR"
    end if
 end function classify

end module
